File: src/python-backend/ml_models/resume_analyzer.py
"""
Resume Analyzer
===============

Analyzes resumes to extract skills, experience, and provide feedback.

Uses:
- spaCy for NER (Named Entity Recognition)
- Custom skill extraction
- ATS (Applicant Tracking System) scoring
- PDF/DOCX parsing

For production:
- Train custom NER model on resume data
- Add industry-specific skill databases
- Implement ATS keyword optimization
"""

import re
from typing import Dict, List, Tuple
import PyPDF2
import docx
from io import BytesIO


class ResumeAnalyzer:
    def __init__(self):
        """Initialize resume analyzer."""
        
        # Skill database (expand this with comprehensive list)
        self.skill_keywords = {
            "programming": [
                "python", "javascript", "java", "c++", "c#", "ruby", "go", "rust",
                "typescript", "php", "swift", "kotlin", "scala", "r"
            ],
            "web": [
                "react", "angular", "vue", "node.js", "express", "django", "flask",
                "fastapi", "next.js", "html", "css", "tailwind", "bootstrap"
            ],
            "database": [
                "sql", "postgresql", "mysql", "mongodb", "redis", "elasticsearch",
                "dynamodb", "cassandra", "firebase", "supabase"
            ],
            "cloud": [
                "aws", "azure", "gcp", "google cloud", "docker", "kubernetes",
                "terraform", "ansible", "jenkins", "ci/cd"
            ],
            "ml_ai": [
                "machine learning", "deep learning", "tensorflow", "pytorch",
                "scikit-learn", "keras", "nlp", "computer vision", "bert", "gpt"
            ],
            "mobile": [
                "react native", "flutter", "ios", "android", "swift", "kotlin"
            ],
            "tools": [
                "git", "github", "gitlab", "jira", "figma", "postman", "vscode"
            ]
        }
        
        # Action verbs for strong resumes
        self.strong_action_verbs = [
            "achieved", "implemented", "led", "developed", "designed", "built",
            "created", "improved", "increased", "reduced", "optimized", "managed",
            "launched", "delivered", "architected", "spearheaded", "initiated"
        ]
        
        # Try to import spaCy (optional)
        try:
            import spacy
            self.nlp = spacy.load("en_core_web_sm")
            self.spacy_available = True
        except:
            self.nlp = None
            self.spacy_available = False
            print("Warning: spaCy not available. Install with: python -m spacy download en_core_web_sm")
    
    def extract_text_from_file(self, file_content: bytes, filename: str) -> str:
        """
        Extract text from PDF or DOCX file.
        
        Args:
            file_content: Binary file content
            filename: Name of the file (to determine type)
        
        Returns:
            Extracted text
        """
        try:
            if filename.lower().endswith('.pdf'):
                return self._extract_from_pdf(file_content)
            elif filename.lower().endswith('.docx'):
                return self._extract_from_docx(file_content)
            else:
                raise ValueError("Unsupported file format. Use PDF or DOCX.")
        except Exception as e:
            raise Exception(f"Error extracting text: {str(e)}")
    
    def _extract_from_pdf(self, content: bytes) -> str:
        """Extract text from PDF."""
        text = ""
        pdf_file = BytesIO(content)
        pdf_reader = PyPDF2.PdfReader(pdf_file)
        
        for page in pdf_reader.pages:
            text += page.extract_text() + "\n"
        
        return text
    
    def _extract_from_docx(self, content: bytes) -> str:
        """Extract text from DOCX."""
        doc_file = BytesIO(content)
        doc = docx.Document(doc_file)
        
        text = ""
        for paragraph in doc.paragraphs:
            text += paragraph.text + "\n"
        
        return text
    
    def extract_skills(self, text: str) -> List[str]:
        """
        Extract skills from resume text.
        
        Args:
            text: Resume text
        
        Returns:
            List of extracted skills
        """
        text_lower = text.lower()
        found_skills = []
        
        # Search for skills from database
        for category, skills in self.skill_keywords.items():
            for skill in skills:
                # Use word boundaries to avoid partial matches
                pattern = r'\b' + re.escape(skill) + r'\b'
                if re.search(pattern, text_lower):
                    # Capitalize properly
                    found_skills.append(skill.title())
        
        # Remove duplicates while preserving order
        seen = set()
        unique_skills = []
        for skill in found_skills:
            if skill.lower() not in seen:
                seen.add(skill.lower())
                unique_skills.append(skill)
        
        return unique_skills
    
    def calculate_ats_score(self, text: str, skills: List[str]) -> int:
        """
        Calculate ATS (Applicant Tracking System) score.
        
        Scoring criteria:
        - Skills mentioned: 30 points
        - Action verbs used: 20 points
        - Quantifiable results: 20 points
        - Proper formatting: 15 points
        - Length (optimal 400-800 words): 15 points
        
        Returns:
            Score out of 100
        """
        score = 0
        text_lower = text.lower()
        
        # 1. Skills (30 points max)
        skills_score = min(len(skills) * 3, 30)
        score += skills_score
        
        # 2. Action verbs (20 points max)
        action_verb_count = sum(
            1 for verb in self.strong_action_verbs
            if verb in text_lower
        )
        action_score = min(action_verb_count * 2, 20)
        score += action_score
        
        # 3. Quantifiable results (20 points max)
        # Look for numbers/percentages
        numbers_count = len(re.findall(r'\d+%|\d+\+|increased|decreased|improved', text_lower))
        quant_score = min(numbers_count * 2, 20)
        score += quant_score
        
        # 4. Formatting (15 points max)
        format_score = 0
        # Check for sections
        if any(section in text_lower for section in ['experience', 'education', 'skills']):
            format_score += 5
        # Check for bullet points or dashes
        if '•' in text or '-' in text:
            format_score += 5
        # Check for email
        if re.search(r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b', text):
            format_score += 5
        score += format_score
        
        # 5. Length (15 points max)
        word_count = len(text.split())
        if 400 <= word_count <= 800:
            length_score = 15
        elif 300 <= word_count < 400 or 800 < word_count <= 1000:
            length_score = 10
        elif 200 <= word_count < 300 or 1000 < word_count <= 1200:
            length_score = 5
        else:
            length_score = 0
        score += length_score
        
        return min(score, 100)
    
    def extract_experience_years(self, text: str) -> int:
        """
        Estimate years of experience from resume.
        
        Args:
            text: Resume text
        
        Returns:
            Estimated years of experience
        """
        # Look for patterns like "2 years", "3+ years", "2019-2023"
        text_lower = text.lower()
        
        # Pattern 1: X years
        years_match = re.findall(r'(\d+)\+?\s*years?', text_lower)
        if years_match:
            return max(int(y) for y in years_match)
        
        # Pattern 2: Date ranges (YYYY-YYYY)
        date_ranges = re.findall(r'(20\d{2})\s*[-–]\s*(20\d{2}|present)', text_lower)
        if date_ranges:
            total_years = 0
            for start, end in date_ranges:
                end_year = 2024 if 'present' in end else int(end)
                start_year = int(start)
                total_years += (end_year - start_year)
            return min(total_years, 20)  # Cap at 20 years
        
        return 0
    
    def generate_strengths(self, text: str, skills: List[str], score: int) -> List[str]:
        """Generate list of resume strengths."""
        strengths = []
        
        if score >= 70:
            strengths.append("Well-structured resume with clear sections")
        
        if len(skills) >= 8:
            strengths.append("Diverse technical skill set")
        
        text_lower = text.lower()
        
        if any(verb in text_lower for verb in self.strong_action_verbs[:5]):
            strengths.append("Uses strong action verbs effectively")
        
        if re.search(r'\d+%|\d+x|increased|improved', text_lower):
            strengths.append("Includes quantifiable achievements")
        
        if 'project' in text_lower:
            strengths.append("Demonstrates hands-on project experience")
        
        if not strengths:
            strengths.append("Resume has basic structure in place")
        
        return strengths[:4]
    
    def generate_improvements(self, text: str, skills: List[str], score: int) -> List[str]:
        """Generate list of resume improvement suggestions."""
        improvements = []
        text_lower = text.lower()
        
        if score < 70:
            improvements.append("Improve overall structure and formatting")
        
        if len(skills) < 6:
            improvements.append("Add more relevant technical skills")
        
        action_verb_count = sum(1 for verb in self.strong_action_verbs if verb in text_lower)
        if action_verb_count < 3:
            improvements.append("Use more action verbs (achieved, implemented, led)")
        
        if not re.search(r'\d+%|\d+x|increased', text_lower):
            improvements.append("Quantify achievements with metrics and numbers")
        
        if 'summary' not in text_lower and 'objective' not in text_lower:
            improvements.append("Add a professional summary section")
        
        if 'certification' not in text_lower and 'certificate' not in text_lower:
            improvements.append("Include relevant certifications if available")
        
        word_count = len(text.split())
        if word_count < 300:
            improvements.append("Expand resume with more details and achievements")
        elif word_count > 1000:
            improvements.append("Reduce length - aim for concise, impactful content")
        
        return improvements[:5]
    
    def suggest_roles(self, skills: List[str]) -> List[str]:
        """Suggest suitable roles based on skills."""
        roles = []
        skills_lower = [s.lower() for s in skills]
        
        # Check for role matches
        if any(skill in skills_lower for skill in ['react', 'angular', 'vue', 'html', 'css']):
            roles.append("Frontend Developer")
        
        if any(skill in skills_lower for skill in ['node.js', 'express', 'django', 'flask', 'fastapi']):
            roles.append("Backend Developer")
        
        if any(skill in skills_lower for skill in ['react', 'node.js']) and \
           any(skill in skills_lower for skill in ['mongodb', 'postgresql', 'mysql']):
            roles.append("Full Stack Developer")
        
        if any(skill in skills_lower for skill in ['tensorflow', 'pytorch', 'machine learning', 'deep learning']):
            roles.append("AI/ML Engineer")
        
        if any(skill in skills_lower for skill in ['python', 'sql', 'pandas', 'scikit-learn']):
            roles.append("Data Scientist")
        
        if any(skill in skills_lower for skill in ['docker', 'kubernetes', 'aws', 'terraform']):
            roles.append("DevOps Engineer")
        
        if any(skill in skills_lower for skill in ['react native', 'flutter', 'swift', 'kotlin']):
            roles.append("Mobile App Developer")
        
        if not roles:
            roles.append("Software Engineer")
        
        return roles[:3]
    
    def analyze_text(self, resume_text: str) -> Dict:
        """
        Analyze resume text and return comprehensive feedback.
        
        Args:
            resume_text: The text content of the resume
        
        Returns:
            Dictionary containing analysis results
        """
        # Extract skills
        skills = self.extract_skills(resume_text)
        
        # Calculate ATS score
        score = self.calculate_ats_score(resume_text, skills)
        
        # Extract experience
        experience_years = self.extract_experience_years(resume_text)
        
        # Generate feedback
        strengths = self.generate_strengths(resume_text, skills, score)
        improvements = self.generate_improvements(resume_text, skills, score)
        
        # Suggest roles
        suggested_roles = self.suggest_roles(skills)
        
        return {
            "score": score,
            "strengths": strengths,
            "improvements": improvements,
            "extracted_skills": skills,
            "experience_years": experience_years,
            "suggested_roles": suggested_roles,
            "word_count": len(resume_text.split()),
            "skill_categories": self._categorize_skills(skills)
        }
    
    def _categorize_skills(self, skills: List[str]) -> Dict[str, List[str]]:
        """Categorize skills by type."""
        categories = {
            "programming": [],
            "web": [],
            "database": [],
            "cloud": [],
            "ml_ai": [],
            "mobile": [],
            "tools": []
        }
        
        skills_lower = [s.lower() for s in skills]
        
        for category, keywords in self.skill_keywords.items():
            for skill in skills:
                if skill.lower() in keywords:
                    categories[category].append(skill)
        
        # Remove empty categories
        return {k: v for k, v in categories.items() if v}


# Example usage
if __name__ == "__main__":
    analyzer = ResumeAnalyzer()
    
    # Test with sample resume text
    sample_resume = """
    John Doe
    john.doe@email.com | +91-9876543210
    
    EXPERIENCE
    Software Engineer at Tech Corp (2021-Present)
    - Developed and implemented React-based web applications, improving user engagement by 35%
    - Built RESTful APIs using Node.js and Express, reducing response time by 40%
    - Managed PostgreSQL databases and optimized queries for better performance
    
    SKILLS
    Programming: Python, JavaScript, TypeScript
    Web: React, Node.js, Express, HTML, CSS
    Database: PostgreSQL, MongoDB
    Tools: Git, Docker, AWS
    
    PROJECTS
    E-commerce Platform: Built a full-stack application using MERN stack
    """
    
    result = analyzer.analyze_text(sample_resume)
    
    print(f"ATS Score: {result['score']}/100")
    print(f"\nSkills Found: {', '.join(result['extracted_skills'])}")
    print(f"\nExperience: {result['experience_years']} years")
    print(f"\nSuggested Roles: {', '.join(result['suggested_roles'])}")
    print(f"\nStrengths:")
    for s in result['strengths']:
        print(f"  - {s}")
    print(f"\nImprovements:")
    for i in result['improvements']:
        print(f"  - {i}")

File: src/lib/ai/resumeAnalyzer.ts
/**
 * Resume Analyzer
 * Skill extraction, ATS scoring, and improvement suggestions
 */

import { tokenize, removeStopWords } from './utils';

export interface ResumeAnalysis {
  score: number;
  strengths: string[];
  improvements: string[];
  extracted_skills: string[];
  experience_years: number;
  suggested_roles: string[];
}

/**
 * Resume Analyzer
 */
export class ResumeAnalyzer {
  private static readonly TECH_SKILLS = new Set([
    'python', 'java', 'javascript', 'typescript', 'c++', 'c#', 'go', 'rust', 'php', 'ruby',
    'react', 'angular', 'vue', 'svelte', 'next.js', 'node.js', 'express', 'django', 'flask',
    'spring', 'laravel', 'rails', 'postgresql', 'mysql', 'mongodb', 'redis', 'elasticsearch',
    'docker', 'kubernetes', 'aws', 'azure', 'gcp', 'terraform', 'jenkins', 'git', 'github',
    'machine learning', 'deep learning', 'tensorflow', 'pytorch', 'scikit-learn', 'pandas',
    'numpy', 'sql', 'nosql', 'rest', 'graphql', 'microservices', 'agile', 'scrum', 'ci/cd',
    'html', 'css', 'sass', 'tailwind', 'bootstrap', 'webpack', 'babel', 'testing', 'jest',
    'linux', 'bash', 'data structures', 'algorithms', 'oop', 'functional programming',
    'firebase', 'supabase', 'netlify', 'vercel', 'heroku', 'digital ocean'
  ]);

  private static readonly SOFT_SKILLS = new Set([
    'leadership', 'communication', 'teamwork', 'problem solving', 'critical thinking',
    'time management', 'creativity', 'adaptability', 'collaboration', 'presentation',
    'negotiation', 'conflict resolution', 'mentoring', 'project management'
  ]);

  private static readonly ACTION_VERBS = new Set([
    'achieved', 'implemented', 'developed', 'created', 'designed', 'built', 'led', 'managed',
    'improved', 'optimized', 'increased', 'reduced', 'launched', 'delivered', 'collaborated',
    'spearheaded', 'initiated', 'coordinated', 'executed', 'established', 'streamlined'
  ]);

  /**
   * Analyze resume and provide comprehensive feedback
   */
  static analyze(resumeText: string): ResumeAnalysis {
    const normalizedText = resumeText.toLowerCase();
    const tokens = tokenize(resumeText);

    // Extract skills
    const extractedSkills = this.extractSkills(normalizedText, tokens);

    // Calculate experience years
    const experienceYears = this.extractExperience(normalizedText);

    // Calculate ATS score
    const score = this.calculateATSScore(normalizedText, tokens, extractedSkills);

    // Identify strengths
    const strengths = this.identifyStrengths(normalizedText, tokens, extractedSkills);

    // Generate improvements
    const improvements = this.generateImprovements(normalizedText, tokens, extractedSkills);

    // Suggest roles
    const suggestedRoles = this.suggestRoles(extractedSkills, experienceYears);

    return {
      score,
      strengths,
      improvements,
      extracted_skills: extractedSkills,
      experience_years: experienceYears,
      suggested_roles: suggestedRoles
    };
  }

  /**
   * Extract technical and soft skills
   */
  private static extractSkills(text: string, tokens: string[]): string[] {
    const foundSkills = new Set<string>();

    // Check for tech skills
    for (const skill of this.TECH_SKILLS) {
      if (text.includes(skill.toLowerCase())) {
        foundSkills.add(this.capitalizeSkill(skill));
      }
    }

    // Check for soft skills
    for (const skill of this.SOFT_SKILLS) {
      if (text.includes(skill.toLowerCase())) {
        foundSkills.add(this.capitalizeSkill(skill));
      }
    }

    // Check for skill section explicitly
    const skillSectionMatch = text.match(/skills?:([^\n]+)/i);
    if (skillSectionMatch) {
      const skillSection = skillSectionMatch[1].toLowerCase();
      const potentialSkills = skillSection.split(/[,;\|]/);
      
      for (const skill of potentialSkills) {
        const cleaned = skill.trim();
        if (cleaned.length > 2) {
          foundSkills.add(this.capitalizeSkill(cleaned));
        }
      }
    }

    return Array.from(foundSkills).slice(0, 15); // Limit to top 15
  }

  /**
   * Extract years of experience
   */
  private static extractExperience(text: string): number {
    // Look for patterns like "2 years", "3+ years", "5-7 years"
    const patterns = [
      /(\d+)\+?\s*years?/i,
      /(\d+)-(\d+)\s*years?/i,
      /experience:\s*(\d+)/i
    ];

    for (const pattern of patterns) {
      const match = text.match(pattern);
      if (match) {
        return parseInt(match[1]);
      }
    }

    // Count number of job entries as fallback
    const jobIndicators = text.match(/\b(internship|intern|job|position|role|engineer|developer|analyst)\b/gi);
    return jobIndicators ? Math.min(Math.floor(jobIndicators.length / 3), 10) : 0;
  }

  /**
   * Calculate ATS (Applicant Tracking System) score
   */
  private static calculateATSScore(text: string, tokens: string[], skills: string[]): number {
    let score = 0;
    const maxScore = 100;

    // 1. Skills coverage (30 points)
    const skillScore = Math.min(30, skills.length * 3);
    score += skillScore;

    // 2. Action verbs (15 points)
    const actionVerbCount = tokens.filter(t => this.ACTION_VERBS.has(t.toLowerCase())).length;
    score += Math.min(15, actionVerbCount * 2);

    // 3. Quantifiable achievements (20 points)
    const numbers = text.match(/\b\d+%?\b/g);
    const quantificationScore = numbers ? Math.min(20, numbers.length * 2) : 0;
    score += quantificationScore;

    // 4. Proper sections (15 points)
    const sections = ['experience', 'education', 'skills', 'projects'];
    const sectionScore = sections.filter(s => text.includes(s.toLowerCase())).length * 3.75;
    score += sectionScore;

    // 5. Length appropriateness (10 points)
    const wordCount = tokens.length;
    const lengthScore = (wordCount >= 200 && wordCount <= 800) ? 10 : 5;
    score += lengthScore;

    // 6. Contact information (10 points)
    const hasEmail = /[\w.-]+@[\w.-]+\.\w+/.test(text);
    const hasPhone = /\b\d{10}\b|\b\d{3}[-.]?\d{3}[-.]?\d{4}\b/.test(text);
    score += hasEmail ? 5 : 0;
    score += hasPhone ? 5 : 0;

    return Math.min(maxScore, Math.round(score));
  }

  /**
   * Identify resume strengths
   */
  private static identifyStrengths(text: string, tokens: string[], skills: string[]): string[] {
    const strengths: string[] = [];

    // Check for diverse skills
    if (skills.length >= 8) {
      strengths.push('Diverse technical skill set');
    }

    // Check for action verbs
    const actionVerbCount = tokens.filter(t => this.ACTION_VERBS.has(t.toLowerCase())).length;
    if (actionVerbCount >= 5) {
      strengths.push('Strong use of action verbs');
    }

    // Check for quantification
    const numbers = text.match(/\b\d+%?\b/g);
    if (numbers && numbers.length >= 5) {
      strengths.push('Good quantification of achievements');
    }

    // Check for projects
    if (text.toLowerCase().includes('project')) {
      strengths.push('Includes relevant projects');
    }

    // Check for certifications
    if (text.toLowerCase().includes('certif')) {
      strengths.push('Professional certifications mentioned');
    }

    // Check for proper structure
    const sections = ['experience', 'education', 'skills'];
    if (sections.every(s => text.toLowerCase().includes(s))) {
      strengths.push('Well-structured with clear sections');
    }

    // Default strength
    if (strengths.length === 0) {
      strengths.push('Basic resume structure present');
    }

    return strengths.slice(0, 5);
  }

  /**
   * Generate improvement suggestions
   */
  private static generateImprovements(text: string, tokens: string[], skills: string[]): string[] {
    const improvements: string[] = [];

    // Check action verbs
    const actionVerbCount = tokens.filter(t => this.ACTION_VERBS.has(t.toLowerCase())).length;
    if (actionVerbCount < 5) {
      improvements.push('Add more action verbs (achieved, implemented, led, etc.)');
    }

    // Check quantification
    const numbers = text.match(/\b\d+%?\b/g);
    if (!numbers || numbers.length < 3) {
      improvements.push('Include metrics and numbers to quantify achievements');
    }

    // Check skills
    if (skills.length < 6) {
      improvements.push('Add more relevant technical skills');
    }

    // Check for summary
    if (!text.toLowerCase().includes('summary') && !text.toLowerCase().includes('objective')) {
      improvements.push('Consider adding a professional summary section');
    }

    // Check for certifications
    if (!text.toLowerCase().includes('certif')) {
      improvements.push('Include relevant certifications if available');
    }

    // Check length
    const wordCount = tokens.length;
    if (wordCount < 200) {
      improvements.push('Resume seems short - add more details about your experience');
    } else if (wordCount > 800) {
      improvements.push('Resume may be too long - focus on most relevant information');
    }

    // Check contact info
    const hasEmail = /[\w.-]+@[\w.-]+\.\w+/.test(text);
    if (!hasEmail) {
      improvements.push('Include professional email address');
    }

    return improvements.slice(0, 5);
  }

  /**
   * Suggest suitable roles based on skills
   */
  private static suggestRoles(skills: string[], experienceYears: number): string[] {
    const skillSet = new Set(skills.map(s => s.toLowerCase()));
    const roles: Array<{ role: string; score: number }> = [];

    // Define role requirements
    const roleRequirements: Record<string, string[]> = {
      'Full Stack Developer': ['javascript', 'react', 'node.js', 'html', 'css'],
      'Frontend Developer': ['react', 'javascript', 'html', 'css', 'typescript'],
      'Backend Developer': ['node.js', 'python', 'java', 'sql', 'api'],
      'Data Scientist': ['python', 'machine learning', 'sql', 'statistics'],
      'ML Engineer': ['python', 'tensorflow', 'pytorch', 'machine learning'],
      'DevOps Engineer': ['docker', 'kubernetes', 'aws', 'ci/cd', 'linux'],
      'Mobile Developer': ['react native', 'flutter', 'swift', 'kotlin'],
      'Software Engineer': ['programming', 'algorithms', 'data structures']
    };

    // Score each role
    for (const [role, requirements] of Object.entries(roleRequirements)) {
      const matchCount = requirements.filter(req => 
        Array.from(skillSet).some(skill => skill.includes(req.toLowerCase()))
      ).length;
      
      const score = matchCount / requirements.length;
      
      if (score > 0.3) { // At least 30% match
        roles.push({ role, score });
      }
    }

    // Sort by score and return top 3
    return roles
      .sort((a, b) => b.score - a.score)
      .slice(0, 3)
      .map(r => r.role);
  }

  /**
   * Capitalize skill name
   */
  private static capitalizeSkill(skill: string): string {
    const specialCases: Record<string, string> = {
      'javascript': 'JavaScript',
      'typescript': 'TypeScript',
      'html': 'HTML',
      'css': 'CSS',
      'sql': 'SQL',
      'nosql': 'NoSQL',
      'api': 'API',
      'rest': 'REST',
      'graphql': 'GraphQL',
      'aws': 'AWS',
      'gcp': 'GCP',
      'ci/cd': 'CI/CD',
      'oop': 'OOP',
      'node.js': 'Node.js',
      'next.js': 'Next.js',
      'react': 'React',
      'vue': 'Vue',
      'angular': 'Angular',
      'docker': 'Docker',
      'kubernetes': 'Kubernetes',
      'mongodb': 'MongoDB',
      'postgresql': 'PostgreSQL',
      'mysql': 'MySQL',
      'firebase': 'Firebase',
      'git': 'Git',
      'github': 'GitHub',
      'python': 'Python',
      'java': 'Java'
    };

    const lower = skill.toLowerCase().trim();
    if (specialCases[lower]) {
      return specialCases[lower];
    }

    return skill.split(' ')
      .map(word => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
      .join(' ');
  }
}
